### PR TITLE
IO related code reorganization

### LINKED ISSUE #46: Organize file IO related functions to their own module
According to our previous discussions, it would make sense to group all of the IO related functions to their own module. Currently `basUtils.py` has most of those functions, but it has also some other functions that would better fit into `atomicUtils.py`.

So the proposal would be to rename `basUtils.py` into `io.py` and move the following functions to `atomicUtils.py`: `findBonds`, `findBondsNP`, `findBonds_`, `findBondsSimple`, `multCell`.

Additionally, `GridUtils.py` also has some IO functions, `saveXSF` etc., which could either be moved to the new `io.py` or at least imported there so that all IO function can be accessed in one place.

But we should wait until #24 gets resolved before doing this.

### PR DESCRIPTION
**Pull request by NikoOinonen**
Fixes #46 

Reorganizing code related to file IO so that all IO functions are in one module called `io`:
- Rename `basUtils` to `io`
- Move some functions from `io` to `atomicUtils`
- Move several functions from `GridUtils` to `io`. Basically `GridUtils` now only functions as an interface to the C++ functions.
- Deleted a couple of functions that were not used anywhere and did not seem important.
- Fix all affected imports.

### CODE REVIEW

**Comment by yakutovicha** [comment]
@NikoOinonen, please resolve the conflicts. I will then review the PR.

File: dev/FARFF/test_FARFF_AuxMaps.py [hunk 1]
```diff
@@ -10,7 +10,6 @@
 #from ppafm import fieldOCL     as FFcl
 #from ppafm import RelaxOpenCL  as oclr
 #from ppafm import common       as PPU
-#from ppafm import basUtils
 #from ppafm.AFMulatorOCL_Simple  import AFMulator
 #from ppafm.GeneratorOCL_Simple2 import InverseAFMtrainer
 #from ppafm.AuxMap import AuxMaps
```
Comment (yakutovicha) [inline dev/FARFF/test_FARFF_AuxMaps.py:15]: Please remove the unnecessary comments and `sys.path.append`. Same for all the occurrences below.

File: dev/bakup/GUImorse.py [hunk 2]
```diff
@@ -394,8 +390,8 @@ def selectMode(self):
         self.loadInputs_New()
         self.getFF()
 
-        #GU.saveVecFieldXsf( "DEBUG_FFcl", self.FF[:,:,:,:3].astype(np.float32), self.lvec )
-        #GU.saveXSF( "DEBUG_FFcl_z.xsf", self.FF[:,:,:,2].astype(np.float32), self.lvec )
+        #io.saveVecFieldXsf( "DEBUG_FFcl", self.FF[:,:,:,:3].astype(np.float32), self.lvec )
+        #io.saveXSF( "DEBUG_FFcl_z.xsf", self.FF[:,:,:,2].astype(np.float32), self.lvec )
```
Comment (yakutovicha) [inline dev/bakup/GUImorse.py:394]:
```suggestion
```

File: dev/bakup/conv_rho_mini.py [hunk 3]
```diff
@@ -14,20 +14,16 @@
 
 import ppafm as PPU
 import ppafm.fieldFFT as fFFT
-import ppafm.GridUtils as GU
-
-#import GridUtils as GU
-#sys.path.append("/u/25/prokoph1/unix/git/ProbeParticleModel")
-
+from ppafm.io import loadXSF, saveXSF
```
Comment (yakutovicha) [inline dev/bakup/conv_rho_mini.py:17]: I would suggest the following pattern everywhere:
```python
from ppafm import io

io.saveXSF(..)  # everywhere
```

File: dev/bakup/conv_rho_mini.py [hunk 4]
```diff
@@ -53,14 +49,14 @@
...
 
 np.save( "E"+namestr+"_mini.npy", Ecut*(PQ*-1.0) )
 np.savez_compressed( "E"+namestr+"_mini.npz", Ecut*(PQ*-1.0) )
 
-#GU.saveXSF( "FF"+namestr+"_x.xsf", Fx*PQ,       lvec1, head=head1 )
-#GU.saveXSF( "FF"+namestr+"_y.xsf", Fy*PQ,       lvec1, head=head1 )
-#GU.saveXSF( "FF"+namestr+"_z.xsf", Fz*PQ,       lvec1, head=head1 )
+#saveXSF( "FF"+namestr+"_x.xsf", Fx*PQ,       lvec1, head=head1 )
+#saveXSF( "FF"+namestr+"_y.xsf", Fy*PQ,       lvec1, head=head1 )
+#saveXSF( "FF"+namestr+"_z.xsf", Fz*PQ,       lvec1, head=head1 )
```
Comment (yakutovicha) [inline dev/bakup/conv_rho_mini.py:61]: eliminate.

```suggestion
```

File: dev/bakup/runOpenCLbatch.py [hunk 5]
```diff
@@ -159,7 +156,7 @@ def evalFF_LJC( iZPP, xyzs, Zs, qs, poss, typeParams, func_runFF=FFcl.runLJC ):
         FF    = evalFFatoms_LJC( atoms, cLJs, poss, func_runFF=FFcl.runLJC )
         FEin =  FF[:,:,:,:4] + Q*FF[:,:,:,4:]
         Tff = time.clock()-t1ff;
-        #GU.saveXSF( geomFileName+'_Fin_z.xsf',  FEin[:,:,:,2], lvec );
+        # io.saveXSF( geomFileName+'_Fin_z.xsf',  FEin[:,:,:,2], lvec );
```
Comment (yakutovicha) [inline dev/bakup/runOpenCLbatch.py:159]: There are many commented lines everywhere in the files you updated. Just remove them, please.

File: plot_results.py [hunk 6]
```diff
@@ -0,0 +1,273 @@
+#!/usr/bin/python -u
```
Comment (yakutovicha) [inline plot_results.py:1]: that file should be gone.

**Review by yakutovicha — CHANGES_REQUESTED** [review]
thanks a lot @NikoOinonen. The two main suggestions I have are:

1. Please always use the same coding pattern:
```python
from ppafm import io

io.loadXYZ(..)
```

2. Please remove unnecessary comments.

File: plot_results.py (hunk 6 above)
Comment (NikoOinonen) [reply plot_results.py]: Deleted.

**Comment by NikoOinonen** [comment]
@yakutovicha Most of the problems are in the scripts inside `dev`, and it looks like most of those scripts are unmaintained and broken. In hindsight, I probably should not have touched them at all, but it was just easier to grep and fix every occurrence of `basUtils` and `GridUtils` than to be selective.

@ProkopHapala Could you check if these scripts are actually needed, and maybe delete or move them to some other branch if they are not important? So that we don't have to spent time modifying them when we do fixes like this.

File: dev/bakup/runOpenCLbatch.py (hunk 5 above)
Comment (ondrejkrejci) [reply dev/bakup/runOpenCLbatch.py:159]: I am having a problem as this reorganization is heavily messing with my pull request, which is to be finished this week (if my son is not ill). Maybe I should wait to this one to be pushed and then I will change the NPY part of it?

**Comment by ondrejkrejci** [comment]
I am having a problem as this reorganization is heavily messing with my pull request, which is to be finished this week (if my son is not ill). Maybe I should wait to this one to be pushed and then I will change the NPY part of it?

**Comment by NikoOinonen** [comment]
@ondrejkrejci We can also merge yours first and then pull those fixes here. Whichever way you think is easiest.

**Comment by ondrejkrejci** [comment]
> @ondrejkrejci We can also merge yours first and then pull those fixes here. Whichever way you think is easiest.

If this is quicker than me, go ahead.

**Comment by yakutovicha** [comment]
> I am having a problem as this reorganization is heavily messing with my pull request, which is to be finished this week (if my son is not ill). Maybe I should wait to this one to be pushed and then I will change the NPY part of it?

My 2 cents on this topic.

An open pull request will have more conflicts with time. That is normal because other people are also working on the code. Therefore, resolving merge conflicts is a natural part of collaborative development. An open pull request should never be a reason for others to wait or to stop working.

**Comment by yakutovicha** [comment]
Blocked till https://github.com/Probe-Particle/ppafm/issues/119 is resolved.

**Comment by yakutovicha** [comment]
@NikoOinonen, please merge the `main` branch into it so the tests are passing.

**Comment by NikoOinonen** [comment]
Merged main and all checks are passing.

**Review by yakutovicha — APPROVED** [review]
Thanks @NikoOinonen. LGTM.

I think it is a good step forward towards high-quality code 👍

**Pull request merged**